src/v0/index.js

```diff
@@ -0,0 +1,227 @@
+
+export default () => {
+  // Returns the assigned nodes (unflattened) for a <content> node.
+  const getAssignedNodes = node => {
+    const slot = node.getAttribute('name');
+
+    let host = node;
+    while (host) {
+      const sr = host[$shadowRoot];
+      if (sr && sr.contains(node)) {
+        break;
+      }
+      host = host.parentNode;
+    }
+
+    if (!host) {
+      return [];
+    }
+
+    let chs = host.childNodes;
+    let chsLen = chs.length;
+
+    const filtered = [];
+
+    for (let a = 0; a < chsLen; a++) {
+      const ch = chs[a];
+      const chSlot = ch.getAttribute ? ch.getAttribute('slot') : null;
+      if (slot === chSlot) {
+        filtered.push(ch);
+      }
+    }
+
+    return filtered;
+  };
+
+  const { getAttribute, setAttribute } = HTMLElement.prototype;
+  const elementInnerHTML = Object.getOwnPropertyDescriptor(Element.prototype, 'innerHTML');
+  const shadowRootInnerHTML = Object.getOwnPropertyDescriptor(ShadowRoot.prototype, 'innerHTML');
+
+  // We do this so creating a <slot> actually creates a <content>.
+  const filterTagName = name => name === 'slot' ? 'content' : name;
+  const createElement = document.createElement.bind(document);
+  const createElementNS = document.createElementNS.bind(document);
+  document.createElement = function(name, ...args) {
+    return createElement(filterTagName(name), ...args);
+  };
+  document.createElementNS = function(name, ...args) {
+    return createElementNS(filterTagName(name), ...args);
+  };
+
+  // Override innerHTML to turn slot nodes into content nodes.
+  function replaceSlotsWithContents(node) {
+    const tree = document.createTreeWalker(node, NodeFilter.SHOW_ELEMENT);
+    const repl = [];
+
+    // Walk the tree and record nods that need replacing.
+    while (tree.nextNode()) {
+      const { currentNode } = tree;
+      if (currentNode.tagName === 'SLOT') {
+        repl.push(currentNode);
+      }
+    }
+
+    repl.forEach(fake => {
+      const name = fake.getAttribute('name');
+      const real = document.createElement('slot');
+      if (name) {
+        real.setAttribute('name', name);
+      }
+
+      fake.parentNode.replaceChild(real, fake);
+      while (fake.hasChildNodes()) {
+        real.appendChild(fake.firstChild);
+      }
+    });
+  }
+  Object.defineProperty(Element.prototype, 'innerHTML', {
+    configurable: true,
+    get: elementInnerHTML.get,
+    set (html) {
+      elementInnerHTML.set.call(this, html);
+      replaceSlotsWithContents(this);
+    },
+  });
+  Object.defineProperty(ShadowRoot.prototype, 'innerHTML', {
+    configurable: true,
+    get: shadowRootInnerHTML.get,
+    set (html) {
+      shadowRootInnerHTML.set.call(this, html);
+      replaceSlotsWithContents(this);
+    },
+  });
+
+
+  // Node
+  // ----
+
+  Object.defineProperty(Node.prototype, 'assignedSlot', {
+    get() {
+      const { parentNode } = this;
+      if (parentNode) {
+        const { shadowRoot } = parentNode;
```

Comment:
This would be a standard node. If it is slot assignable (light DOM) it's parent will be the host and should have a `shadowRoot` property. Shadow roots cannot be detached from a host and slots cannot work when outside of a shadow root. If the `shadowRoot` property doesn't exist, it means that it's "closed" and it should return null.